### ISSUE TITLE
json dumped datetime behaviour change

### ISSUE THREAD

**Issue by sidekick-eimantas** [pydantic/pydantic#7035]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

pydantic v1 .json() would dump datetimes as follows:
```py
>>> a.json()
'{"time": "2023-08-08T08:02:10.013540+00:00"}'
```

pydantic v2 .json() and .model_dump_json() dumps datetimes as follows:
```py
>>> a.model_dump_json()
'{"time":"2023-08-08T07:57:57.786989Z"}'
```

The new format cannot be loaded using `datetime.datetime.fromisoformat`

Is there a way to instruct pydantic to dump datetimes the v1 way?

### Python, Pydantic & OS Version

```Text
pydantic version: 2.1.1
        pydantic-core version: 2.4.0
          pydantic-core build: profile=release pgo=false mimalloc=true
                 install path: /Users/Eimantas/git/skm/wg/.venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.7 (main, Mar 13 2023, 10:00:01) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.4.1-arm64-arm-64bit
     optional deps. installed: ['typing-extensions']
```

Selected Assignee: @hramezani

**Comment by hramezani** [comment]
Thanks @sidekick-eimantas for reporting this 🙏 

You can use [field_serializer](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.field_serializer) to customize your serialization:

```py
from datetime import datetime

from pydantic import BaseModel, field_serializer


class Model(BaseModel):
    time: datetime

    @field_serializer('time')
    def customise_time_serialization(self, v, info) -> str:
        return v.isoformat()


m = Model(time=datetime.now())
print(m.model_dump_json())  # {"time":"2023-08-08T11:36:42.220234"}
```

**Comment by MrNaif2018** [comment]
@hramezani Hi! Sorry for hijacking. Is there any way to alter this behaviour globally? (same story for Decimal/float serialization) Because e.g. in my fastapi app sometimes I can't define exact model for dynamic fields returned (it is type hinted as just dict), and I get incompatible datetime returned

**Comment by hramezani** [comment]
You can define your custom type and use it in all places instead of adding `field_serializer` on every model:

```py
from datetime import datetime

from pydantic import BaseModel, PlainSerializer

from typing_extensions import Annotated


MyDateTime = Annotated[datetime, PlainSerializer(lambda x: x.isoformat())]


class Model(BaseModel):
    time: MyDateTime


m = Model(time=datetime.now())
print(m.model_dump_json())  # {"time":"2023-08-08T11:36:42.220234"}
```

**Comment by MrNaif2018** [comment]
But what if I have a model like

```
class Model(BaseModel):
      custom_data: dict
```
      
And this dict (contains dynamic data added by plugins) contains a datetime object?

**Comment by hramezani** [comment]
Then you may need to add a `field_serializer` for `custom_data` and then serialize datetime values differently.

You can find more information about customize serialization in the [doc](https://docs.pydantic.dev/2.1/usage/serialization/#custom-serializers).

**Comment by slafs** [comment]
Oh, this took me by surprise TBH.

What's interesting is that it seems this format is only applied to UTC datetimes, any other TZ is serialized like in v1.

So e.g. this piece (wrapped with a small FastAPI app, sorry):
```py
import zoneinfo
from datetime import datetime

from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel

app = FastAPI()

warsaw_tz = zoneinfo.ZoneInfo("Europe/Warsaw")
utc_tz = zoneinfo.ZoneInfo("UTC")


class Model(BaseModel):
    dt1: datetime
    dt2: datetime


warsaw_dt = datetime(2023, 9, 5, 8, 38, 15, 886840, tzinfo=warsaw_tz)
utc_dt = datetime(2023, 9, 5, 8, 38, 15, 886840, tzinfo=utc_tz)
m = Model(dt1=warsaw_dt, dt2=utc_dt)


@app.get("/")
def root() -> Model:
    return m


client = TestClient(app)
res = client.get("/")
print(res.text)
```
on v1 prints:
```json
{
  "dt1": "2023-09-05T08:38:15.886840+02:00",
  "dt2": "2023-09-05T08:38:15.886840+00:00"
}
```
while on v2 it gives:
```json
{
  "dt1": "2023-09-05T08:38:15.886840+02:00",
  "dt2": "2023-09-05T08:38:15.886840Z"
}
```

Correct me if I'm wrong, but this feels to me like an oversight. Can we at least agree that this behaviour is (a) inconsistent and (b) undocumented (I couldn't find anything in the migration guide nor serialization docs)?

So, while I understand I can fix/customise this per field, it seems like a lot of work (if one has many models) that could be avoided 😬.

**Comment by slafs** [comment]
OK, after reading a bit more about the standard, saying this behaviour is "inconsitent" might be an exaggeration 😅.

For reference:
- https://en.wikipedia.org/wiki/ISO_8601#Time_zone_designators
- https://github.com/python/cpython/issues/90772

**Comment by thibaultbetremieux** [comment]
Hi, I'm having the same problem but in my case it's not that easy to fix, since I have two base models that I inherit and in many children there are datetime fields.

I can see in my tests after migrating to v2 that the datetime format has changed.

There is of course the approach of changing the datetime field everywhere as you suggested (either with a custom type or adding a validation method each time), but I was hoping for a less cumbersome and more global approach.

I'm sorry if this is a stupid question but why isn't there something like `model_config = ConfigDict(ser_json_datetime='iso8601')`? We indeed already have `ser_json_timedelta` for changing the output of all timedeltas.

**Comment by RollerKnobster** [comment]
I agree with @thibaultbetremieux, I would love to have a config handle to choose between datetime serialization behavious, especially since Zulu timezone format is (to my knowledge) not yet natively supported by Python's `datetime` anyways.
